### PR TITLE
PR2: Contracts Client Controllers (Server + to Chain Initialization)

### PR DESCRIPTION
**Pull request by dantrinh**
Make Contracts Controller to handle the connection toggling between contracts.

Did not address this because it didn't make sense: Fix the ModelFromJson() for Chain to not use the Contracts[0] address

references in AAVE, Compound, NFTAdapter, and TokenAdapter controllers fetch correct contract

## Does this PR affect any server routes?
- [ ] yes
- [ ] no

## Have proper tags been added (for bug, enhancement, breaking change)?
- [ ] yes

### CODE REVIEW

File: packages/commonwealth/client/scripts/app.ts [hunk 1]
```diff
@@ -67,7 +67,13 @@ export async function initAppState(
         data.chains
           .filter((chain) => chain.active)
           .map((chain) => {
-            delete chain.ChainNode;
```
Comment (zakhap) [inline packages/commonwealth/client/scripts/app.ts:70]: still need this line

Comment (dantrinh) [reply packages/commonwealth/client/scripts/app.ts:70]: accidental delete! adding back

File: packages/commonwealth/client/scripts/controllers/chain/ethereum/contracts/adapter.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import { IChainAdapter, ChainEntity, ChainEvent, NodeInfo, ChainInfo } from 'models';
+
+import ChainEntityController from 'controllers/server/chain_entities';
+import { IApp } from 'state';
+
+import { notifyError } from 'controllers/app/notifications';
+import { AaveTypes } from 'chain-events/src';
+import AaveChain from '../aave/chain';
+
+export default class Aave extends IChainAdapter<EthereumCoin, EthereumAccount> {
```
Comment (zakhap) [inline packages/commonwealth/client/scripts/controllers/chain/ethereum/contracts/adapter.ts:15]: What is this? Can we delete this file?

Comment (dantrinh) [reply packages/commonwealth/client/scripts/controllers/chain/ethereum/contracts/adapter.ts:15]: Yes can delete this, I'll delete!

Comment (dantrinh) [reply packages/commonwealth/client/scripts/controllers/chain/ethereum/contracts/adapter.ts:15]: this was a POC of a design that we eventually rejected

**Review by zakhap — CHANGES_REQUESTED** [review]
Left some comments on things that should be changed! Let me know if you have any questions 🚀

File: packages/commonwealth/client/scripts/controllers/server/contracts.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        return this._store.getById(id);
+    }
+    public getByType(type: string) {
+        return this._store.getContractByType(type);
+    }
+    public addToStore(contract: Contract) {
+        return this._store.add(contract);
+    }
+
+    // TODO: Add a method to add a contract via post request
```
[the remark this answers is not in the document]
Comment (dantrinh) [reply packages/commonwealth/client/scripts/controllers/server/contracts.ts:33]: yes!

File: packages/commonwealth/client/scripts/stores/ContractsStore.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import IdStore from './IdStore';
+import { Contract } from '../models';
+
+// Models a store of all the contracts
+class ContractsStore extends IdStore<Contract> {
+
```
[the remark this answers is not in the document]
Comment (dantrinh) [reply packages/commonwealth/client/scripts/stores/ContractsStore.ts:6]: ah lol, I did not realize I am not storing them and am just referencing all of them each time

[the remark this answers is not in the document]
Comment (dantrinh) [reply packages/commonwealth/client/scripts/stores/ContractsStore.ts:6]: Will add a store array like you propose!

File: packages/commonwealth/client/scripts/controllers/chain/ethereum/NftAdapter.ts [hunk 5]
```diff
@@ -31,12 +31,18 @@ export default class Nft extends Ethereum implements ITokenAdapter {
   // Extensions of Ethereum
   constructor(meta: ChainInfo, app: IApp) {
     super(meta, app);
-    this.contractAddress = meta.address;
+    // iterate through selectedChain.Contracts for the Nft type and return the address
+    const nftContracts = app.contracts.getByType('erc721').filter((c) => c.symbol === meta.default_symbol);
```
[the remark this answers is not in the document]
Comment (dantrinh) [reply packages/commonwealth/client/scripts/controllers/chain/ethereum/NftAdapter.ts:35]: I forgot why I did this but the logic here is kinda funky. I am going to filter by tokenName rather and match on that

File: packages/commonwealth/client/scripts/controllers/chain/ethereum/aave/chain.ts [hunk 6]
```diff
@@ -17,9 +17,15 @@ export default class AaveChain extends EthereumChain {
   public async init(selectedChain: ChainInfo) {
     await super.resetApi(selectedChain);
     await super.initMetadata();
+    // iterate through selectedChain.Contracts for the Aave type and return the address
+    const aaveContracts = this.app.contracts.getByType('aave').filter((c) => c.symbol === selectedChain.default_symbol);
```
[the remark this answers is not in the document]
Comment (dantrinh) [reply packages/commonwealth/client/scripts/controllers/chain/ethereum/aave/chain.ts:21]: same here going to match by tokenName rather

**Review by zakhap — COMMENTED** [review]
Seems good for me, but i'm gonna deploy on staging and QA for a bit before merging

**Review by zakhap — APPROVED** [review]

File: packages/commonwealth/client/scripts/app.ts [hunk 7]
```diff
@@ -599,20 +606,24 @@ Promise.all([$.ready, $.get('/api/domain')]).then(
...
-            const chain = chains.find((o) => o.address === scope);
-            if (chain) {
-              const pagePath = window.location.href.substr(
-                window.location.href.indexOf(scope) + scope.length
-              );
-              m.route.set(`/${chain.id}${pagePath}`);
-            }
-          }
-        }
+        /* Deprecating the contract_address => community pipeline.
```
Comment (jnaviask) [inline packages/commonwealth/client/scripts/app.ts:609]: based

File: packages/commonwealth/client/scripts/controllers/chain/ethereum/NftAdapter.ts [hunk 8]
```diff
@@ -36,16 +36,18 @@ export default class Nft extends Ethereum implements ITokenAdapter {
   // Extensions of Ethereum
   constructor(meta: ChainInfo, app: IApp) {
     super(meta, app);
-    this.contractAddress = meta.address;
+    // iterate through selectedChain.Contracts for the Nft type and return the address
+    const nftContracts = app.contracts.getByType('erc721');
```
Comment (jnaviask) [inline packages/commonwealth/client/scripts/controllers/chain/ethereum/NftAdapter.ts:40]: can we have a list or enum of types somewhere

File: packages/commonwealth/client/scripts/controllers/server/contracts.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
+// Create controller for contract model
+//
+// Language: typescript
```
Comment (jnaviask) [inline packages/commonwealth/client/scripts/controllers/server/contracts.ts:3]: what's this header about?

**Review by jnaviask — COMMENTED** [review]
View changes only? Makes sense to me. Left a few small comments otherwise good to go.

File: packages/commonwealth/client/scripts/app.ts (hunk 7 above)
Comment (zakhap) [reply packages/commonwealth/client/scripts/app.ts:609]: 💯

File: packages/commonwealth/client/scripts/app.ts [hunk 10]
```diff
@@ -67,7 +67,14 @@ export async function initAppState(
           .filter((chain) => chain.active)
           .map((chain) => {
             delete chain.ChainNode;
-            // add chain.Contracts to ContractsController here (can be one at a time, in the loop like below)
+            chain.Contracts.map((contract) => {
+              return app.contracts.addToStore(
+                Contract.fromJSON({
+                  ...contract,
```
[the remark this answers is not in the document]
Comment (zakhap) [reply packages/commonwealth/client/scripts/app.ts:73]: what do you mean by this?

File: packages/commonwealth/client/scripts/controllers/chain/ethereum/NftAdapter.ts (hunk 8 above)
Comment (zakhap) [reply packages/commonwealth/client/scripts/controllers/chain/ethereum/NftAdapter.ts:40]: 👍

**Review by jnaviask — APPROVED** [review]

**Review by zakhap — CHANGES_REQUESTED** [review]
Screen Shot 2022-10-12 at 1 01 58 PM

critical bug... we're storing all the contracts from our entire db in the contract controller, and then when filtering by type, for example this dydx page, we return 2, not just the one for the community.: 
Screen Shot 2022-10-12 at 1 15 14 PM

TODO:
- [ ] Contracts Controller should sort by community + type.

**Review by jnaviask — CHANGES_REQUESTED** [review]
See Zak's QA bug

**Comment by dantrinh** [comment]
Screen Shot 2022-10-12 at 11 30 50 AM

**Comment by dantrinh** [comment]
> Screen Shot 2022-10-12 at 11 30 50 AM

There seems to be an issue with the migration. Upon reseting the db and attempting to migrate

**Review by dantrinh — COMMENTED** [review]
Now LGTM @zakhap

**Review by zakhap — APPROVED** [review]
Everything seems to still work, LGTM!

**Review by jnaviask — APPROVED** [review]

**Pull request merged**